### PR TITLE
ddl: disallow generated column referring to auto_increment column

### PR DESCRIPTION
**Pull request by spongedu**
### What problem does this PR solve? 
As is described in [CREATE TABLE and Generated Columns](https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html) and [ALTER TABLE and Generated Columns](https://dev.mysql.com/doc/refman/5.7/en/alter-table-generated-columns.html), `generated` columns should not refer to any `auto_increment` columns.

For example, in MySQL:
```sql
mysql> create table t1 (a bigint not null primary key auto_increment, b bigint as (a + 1));
ERROR 3109 (HY000): Generated column 'b' cannot refer to auto-increment column.
mysql> create table t (a bigint not null primary key auto_increment, b bigint as (a + 1));
ERROR 3109 (HY000): Generated column 'b' cannot refer to auto-increment column.
mysql> create table t ( a int  not null primary key auto_increment, b int);
Query OK, 0 rows affected (0.03 sec)

mysql> ALTER TABLE t ADD COLUMN c2 INT GENERATED ALWAYS AS (a + 1);
ERROR 3109 (HY000): Generated column 'c2' cannot refer to auto-increment column.
mysql> ALTER TABLE t ADD COLUMN c INT GENERATED ALWAYS AS (b + 1);
Query OK, 0 rows affected (0.03 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> ALTER TABLE t MODIFY COLUMN c int GENERATED ALWAYS AS (a + 5);
ERROR 3109 (HY000): Generated column 'c' cannot refer to auto-increment column.
```

### What is changed and how it works?
Add related check on ddl.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

Related changes

This pr depends on related error definition, which refers to another pr in [parser](https://github.com/pingcap/parser/pull/194).
After [parser related changes](https://github.com/pingcap/parser/pull/194) has been merged, the modification on `go.mod` and `go.sum` should be reverted.

### CODE REVIEW

**Comment by spongedu** [comment]
@zz-jason @winkyao  PTAL

File: ddl/generated_column.go [hunk 1]
```diff
@@ -153,3 +153,14 @@ func checkModifyGeneratedColumn(originCols []*table.Column, oldCol, newCol *tabl
 	}
 	return nil
 }
+
+// checkAutoIncrementRef checks if an generated column depends on an auto-increment column and raise an error if so.
```
Comment (zimulala) [inline ddl/generated_column.go]: s/`raise`/`raises`

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -639,10 +639,23 @@ func checkDuplicateColumn(cols []interface{}) error {
...
+		if option.Tp == ast.ColumnOptionAutoIncrement {
+			return true
+		}
+	}
+	return false
+}
+
 func checkGeneratedColumn(colDefs []*ast.ColumnDef) error {
 	var colName2Generation = make(map[string]columnGenerationInDDL, len(colDefs))
+	autoIncrementColumns := make(map[string]struct{})
```
Comment (zimulala) [inline ddl/ddl_api.go]: It seems that only one AUTO_INCREMENT column per table, could we use `autoIncrementColumns` directly?

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -2171,6 +2206,25 @@ func (d *ddl) ModifyColumn(ctx sessionctx.Context, ident ast.Ident, spec *ast.Al
...
+	for _, option := range specNewColumn.Options {
+		if option.Tp == ast.ColumnOptionGenerated {
+			_, t, err := d.getSchemaAndTableByIdent(ctx, ident)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			autoIncremtnColumns := collectAutoIncrementColumns(t.Meta())
+			referableColNames := make(map[string]struct{}, len(t.Cols()))
+			for _, col := range t.Cols() {
+				referableColNames[col.Name.L] = struct{}{}
```
Comment (zimulala) [inline ddl/ddl_api.go]: `referableColNames` doesn't be used? Please remove it.

File: ddl/ddl_api.go (hunk 2 above)
Comment (spongedu) [reply ddl/ddl_api.go]: yep, seems so~

**Comment by spongedu** [comment]
@zimulala PTAL

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -639,9 +639,23 @@ func checkDuplicateColumn(cols []interface{}) error {
...
+	}
+	return false
+}
+
 func checkGeneratedColumn(colDefs []*ast.ColumnDef) error {
 	var colName2Generation = make(map[string]columnGenerationInDDL, len(colDefs))
+	var exists bool
+	var autoIncrementColumn string
 	for i, colDef := range colDefs {
+		if !exists && checkIsAutoIncrementColumn(colDef){
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: `exists` is redundant

```
if autoIncrementColumn == "" {
    if checkIsAutoIncrementColumn(colDef){
        autoIncrementColumn = colDef.Name.Name.L
    }
}
```

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -1359,13 +1384,23 @@ func handleTableOptions(options []*ast.TableOption, tbInfo *model.TableInfo) err
...
 		if mysql.HasAutoIncrementFlag(col.Flag) {
-			return true
+			return true, col.Name.L
 		}
 	}
-	return false
+	return false, ""
+}
+
+func collectAutoIncrementColumns(tbInfo *model.TableInfo) map[string]struct{} {
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: This function is not used?

File: ddl/ddl.go [hunk 6]
```diff
@@ -121,8 +121,9 @@ var (
 	// errJSONUsedAsKey forbiddens to use JSON as key or index.
 	errJSONUsedAsKey = terror.ClassDDL.New(codeJSONUsedAsKey, mysql.MySQLErrName[mysql.ErrJSONUsedAsKey])
 	// errBlobCantHaveDefault forbiddens to give not null default value to TEXT/BLOB/JSON.
-	errBlobCantHaveDefault = terror.ClassDDL.New(codeBlobCantHaveDefault, mysql.MySQLErrName[mysql.ErrBlobCantHaveDefault])
-	errTooLongIndexComment = terror.ClassDDL.New(codeErrTooLongIndexComment, mysql.MySQLErrName[mysql.ErrTooLongIndexComment])
+	errBlobCantHaveDefault       = terror.ClassDDL.New(codeBlobCantHaveDefault, mysql.MySQLErrName[mysql.ErrBlobCantHaveDefault])
+	errTooLongIndexComment       = terror.ClassDDL.New(codeErrTooLongIndexComment, mysql.MySQLErrName[mysql.ErrTooLongIndexComment])
+	ErrGeneratedColumnRefAutoInc = terror.ClassDDL.New(codeErrGeneratedColumnRefAutoInc, "Generated column '%s' cannot refer to auto-increment column.")
```
Comment (lamxTyler) [inline ddl/ddl.go]: What abount `mysql.MySQLErrName[mysql.ErrGeneratedColumnRefAutoInc]`?

File: ddl/ddl_api.go (hunk 5 above)
Comment (spongedu) [reply ddl/ddl_api.go]: yes, this function is useless after some modifications. I'll remove this. thx

File: ddl/ddl_api.go (hunk 4 above)
Comment (spongedu) [reply ddl/ddl_api.go]: ok

File: ddl/ddl.go (hunk 6 above)
Comment (spongedu) [reply ddl/ddl.go]: ok

**Comment by spongedu** [comment]
@tiancaiamao @lamxTyler  PTAL :) thx~

**Comment by tiancaiamao** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @crazycs520 @zimulala

**Comment by spongedu** [comment]
Since https://github.com/pingcap/parser/pull/194 has been merged, I've fixed related changes in `go.mod`

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -685,6 +699,16 @@ func checkGeneratedColumn(colDefs []*ast.ColumnDef) error {
 			}
 		}
 	}
+
+	// Check whether the generated column refers to any auto-increment columns
+	if exists {
+		for colName, generated := range colName2Generation {
+			if _, found := generated.dependences[autoIncrementColumn]; found {
+				return errors.Trace(ErrGeneratedColumnRefAutoInc.GenWithStackByArgs(colName))
```
Comment (zz-jason) [inline ddl/ddl_api.go]: no need to call `errors.Trace()` since `GenWithStackByArgs()` already generates the call stack.

Comment (spongedu) [reply ddl/ddl_api.go]: ok

Comment (spongedu) [reply ddl/ddl_api.go]: @zz-jason PTAL

**Comment by tiancaiamao** [comment]
Conflict again...

**Comment by zz-jason** [comment]
/run-all-tests tidb-test=release-2.1 tikv=release-2.1 pd=release-2.1

**Pull request merged**